### PR TITLE
Feature/89 trace summary in html report

### PR DESCRIPTION
**Pull request by redcatbear**
## Features
* #89: HTML tracing report now has summary and ToC.

### CODE REVIEW

File: src/main/java/org/itsallcode/openfasttrace/report/view/AbstractStreamableViewContainer.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import java.io.PrintStream;
+
+/**
+ * Abstract base class for View Containers that can be rendered to an output
+ * stream.
+ */
+public abstract class AbstractStreamableViewContainer extends AbstractViewContainer
+{
+    protected PrintStream stream;
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/openfasttrace/report/view/AbstractStreamableViewContainer.java]:
```suggestion
    protected final PrintStream stream;
```

File: src/main/java/org/itsallcode/openfasttrace/report/view/AbstractStreamableViewContainer.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    public AbstractStreamableViewContainer(final PrintStream stream)
+    {
+        super();
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/openfasttrace/report/view/AbstractStreamableViewContainer.java]:
```suggestion
```

File: src/main/java/org/itsallcode/openfasttrace/report/view/AbstractViewContainer.java [hunk 3]
```diff
@@ -31,20 +31,49 @@
...
 
-    /*
-     * (non-Javadoc)
+    /**
+     * Create a new instance of type {@link AbstractViewContainer}.
      * 
-     * @see org.itsallcode.openfasttrace.report.view.Viewable#render(int)
+     * @param id
+     *            unique container ID
+     * @param title
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/openfasttrace/report/view/AbstractViewContainer.java]:
```suggestion
     * @param title container title
```

File: src/main/java/org/itsallcode/openfasttrace/report/view/AbstractViewContainer.java [hunk 4]
```diff
@@ -31,20 +31,49 @@
...
+    protected final String id;
+    protected final String title;
 
     /**
      * Create a new instance of type {@link AbstractViewContainer}.
      */
     public AbstractViewContainer()
     {
-        this.children = new ArrayList<>();
+        this(null, null);
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/openfasttrace/report/view/AbstractViewContainer.java]: If id can be null: maybe make it Optional<String>?

File: src/main/java/org/itsallcode/openfasttrace/report/view/IndentationHelper.java [hunk 5]
```diff
@@ -34,8 +34,15 @@ private IndentationHelper()
         // prevent instantiation.
     }
 
+    /**
+     * Create indentation prefix (i.e. white spaces)
+     * 
+     * @param level
+     *            indentation level
+     * @return white spaces
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/openfasttrace/report/view/IndentationHelper.java]:
```suggestion
     * @return <code>level</code> white spaces
```

File: src/main/java/org/itsallcode/openfasttrace/report/view/html/HtmlReportSummary.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+     * @param stream
+     *            print stream
+     */
+    public HtmlReportSummary(final PrintStream stream)
+    {
+        super(stream);
+    }
+
+    @Override
+    protected void renderBeforeChildren(final int level)
```
Comment (kaklakariada) [inline src/main/java/org/itsallcode/openfasttrace/report/view/html/HtmlReportSummary.java]: Idea for refactoring: indentation is a cross cutting concern, maybe extract it into a base class. Maybe ViewContainer return HtmlElement objects with element name, attributes and children that are then rendered?

File: src/test/java/org/itsallcode/openfasttrace/report/view/html/TestHtmlReportDetails.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+import org.itsallcode.openfasttrace.report.view.Viewable;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class TestHtmlReportDetails extends AbstractTestHtmlRenderer
+{
+    @Override
+    @BeforeEach
+    public void prepareEachTest()
```
Comment (kaklakariada) [inline src/test/java/org/itsallcode/openfasttrace/report/view/html/TestHtmlReportDetails.java]: Is this necessary? The parent class method should be automatically called?

File: src/test/java/org/itsallcode/openfasttrace/report/view/html/TestHtmlTraceSummary.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    private void renderTaceSummaryOnIndentationLevel(final int indentationLevel)
+    {
+        final Viewable view = this.factory.createTraceSummary(this.traceMock);
+        view.render(indentationLevel);
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = { 0, 1, 50, 99 })
```
Comment (kaklakariada) [inline src/test/java/org/itsallcode/openfasttrace/report/view/html/TestHtmlTraceSummary.java]: 👍

File: src/main/java/org/itsallcode/openfasttrace/report/view/AbstractViewContainer.java (hunk 4 above)
Comment (redcatbear) [reply src/main/java/org/itsallcode/openfasttrace/report/view/AbstractViewContainer.java]: If it was part of the external API, yes. But `null` does not appear on the outside, so I think this is okay.

File: src/test/java/org/itsallcode/openfasttrace/report/view/html/TestHtmlReportDetails.java (hunk 7 above)
Comment (redcatbear) [reply src/test/java/org/itsallcode/openfasttrace/report/view/html/TestHtmlReportDetails.java]: That was done by automatic refactoring. Good that you saw it, I completely missed that part.

File: src/main/java/org/itsallcode/openfasttrace/report/view/html/HtmlReportSummary.java (hunk 6 above)
Comment (redcatbear) [reply src/main/java/org/itsallcode/openfasttrace/report/view/html/HtmlReportSummary.java]: I think in the long run we will go for a template solution anyway. I would rather invest the time into that.

**Pull request merged**